feat: Add filters for table calculations

Shifan-Gu opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

Closes: [7480](https://github.com/lightdash/lightdash/issues/7480) and [1270](https://github.com/lightdash/lightdash/issues/1270)

### Description:
allow add filter based on table calculations.

jaffle shop example:

step 1: create a ranking table calculation using total order amount
<img width="791" alt="ranking_table_calculation" src="https://github.com/lightdash/lightdash/assets/101873365/d1f4f88a-844e-49e5-b200-73e6cde9f522">

step 2:
<img width="401" alt="select_table_calculation" src="https://github.com/lightdash/lightdash/assets/101873365/2d3379b7-427c-4bd3-91a7-5ce022323c3e">


step 3:
create filter to find top 5 most expensive order
<img width="1314" alt="top_5_order" src="https://github.com/lightdash/lightdash/assets/101873365/12b7f3d4-49a1-4b02-a3f6-31083dafc428">


### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

TuringLovesDeathMetal commented:
one thing worth checking here for whoever is reviewing: what happens when the table calculation returns a string or boolean data type

TuringLovesDeathMetal commented:
Also, @Shifan-Gu the engineering team is at an offsite this week, so we won't review it this week, but we'll take a look at it once the engineers are back! 👀

rephus commented on packages/backend/src/models/DashboardModel/DashboardModel.ts:
```diff
@@ -551,6 +559,9 @@ export class DashboardModel {
             );
 
+        const tableCalculationFilters = view?.filters?.tableCalculations;
+        view.filters.tableCalculations = tableCalculationFilters || [];
```
Is this necessary ? It feels a bit redundant. I don't see it happening for other filters like `dimensions` or `metrics`

Shifan-Gu replied:
Hi Rephus, reason doing so is because dashbaord view is stored as json in the database, so i'm not sure what's the recommended way to migrate the json object to include default table calculation filters list, so instead i set it to empty list here if its not exist in the database.

rephus commented on packages/common/src/compiler/filtersCompiler.ts:
```diff
@@ -299,6 +301,24 @@ const renderBooleanFilterSql = (
+        case TableCalculationFormatType.DEFAULT:
+        case TableCalculationFormatType.PERCENT:
+        case TableCalculationFormatType.CURRENCY:
+        case TableCalculationFormatType.NUMBER: {
```
does that mean we can only filter numeric results from table calculations ?

rephus commented on packages/backend/src/queryBuilder.ts:
```diff
@@ -720,9 +735,12 @@ export const buildQuery = ({
             : '';
         const secondQuery = [finalSelect, finalFrom, finalSqlWhere].join('\n');
+        const finalQuery = tableCalculationFilters
+            ? `SELECT * FROM (${secondQuery}) query_result WHERE ${tableCalculationFilters}`
```
note to self: test this

rephus commented on packages/backend/src/queryBuilder.ts:
```diff
@@ -604,6 +605,14 @@ export const buildQuery = ({
+        if (fieldType === FieldType.TABLE_CALCULATION) {
+            const field = compiledMetricQuery.compiledTableCalculations?.find(
+                (tc) =>
+                    `table_calculation_${tc.name}` === filter.target.fieldId,
```
note to self: test this

rephus commented on packages/common/src/types/field.ts:
```diff
@@ -123,6 +123,15 @@ export type TableCalculation = {
     format?: TableCalculationFormat;
 };
 
+export interface TableCalculationFilter extends Field {
```
I don't think TableCalculationFilter should to extend Field, they are different types, a filter should not be a field.

rephus replied:
WHy do we need this type ? FilterableItem already supports TableCalculation

Shifan-Gu replied:
it conflicts some where in the code i remember if i only use FilterableItem in the code, let me figure this out and come back to you.

rephus commented on packages/common/src/utils/filters.ts:
```diff
@@ -67,8 +70,8 @@ export const hasNestedGroups = (filters: Filters): boolean => {
     return (
-        (!!filters.dimensions && hasGroups(filters.dimensions)) ||
-        (!!filters.metrics && hasGroups(filters.metrics))
+        (!!filters.metrics && hasGroups(filters.metrics)) ||
```
why have we removed the `(!!filters.dimensions && hasGroups(filters.dimensions))` result ?

Shifan-Gu replied:
good catch, that's a mistake when resolving the conflict.

rephus commented on packages/frontend/src/hooks/dashboard/useDashboardFiltersForExplore.ts:
```diff
@@ -34,6 +34,14 @@ const useDashboardFiltersForExplore = (
                 ...dashboardFilters.metrics,
                 ...(dashboardTemporaryFilters?.metrics ?? []),
             ]),
+            tableCalculations: getDashboardFilterRulesForTile(
```
This method does not exist, perhaps you mean `getDashboardFilterRulesForTileAndTables` ?

Shifan-Gu replied:
same as above, i will correct this

rephus commented:
2 compilation errors: 

```
18.23 src/hooks/dashboard/useDashboardFiltersForTile.ts(14,5): error TS2741: Property 'tableCalculations' is missing in type '{ dimensions: DashboardFilterRule<ConditionalOperator, DashboardFieldTarget, any, any>[]; metrics: DashboardFilterRule<ConditionalOperator, DashboardFieldTarget, any, any>[]; }' but required in type 'DashboardFilters'.
18.23 src/utils/dateFilter.ts(52,5): error TS2741: Property 'tableCalculations' is missing in type '{ dimensions: DashboardFilterRule<ConditionalOperator, DashboardFieldTarget, any, any>[]; metrics: DashboardFilterRule<ConditionalOperator, DashboardFieldTarget, any, any>[]; }' but required in type 'DashboardFilters'.
```

rephus commented:
There are some APi e2e errors

>     1) Should get success response (200) from POST chart results with filters

rephus commented:
Using a more complex TC expression (just adding brackets) does not work 

![Screenshot from 2023-11-20 10-39-15](https://github.com/lightdash/lightdash/assets/1983672/a9e0334a-aa0c-4a88-a831-468184c6bfbf)

![Screenshot from 2023-11-20 10-39-08](https://github.com/lightdash/lightdash/assets/1983672/5dab4e57-8077-46c6-b6e1-1438ec5d9822)

![Screenshot from 2023-11-20 10-38-59](https://github.com/lightdash/lightdash/assets/1983672/38823079-3b0c-4e9e-b697-c7bfe7c0269e)

rephus commented:
Returning a string (concat) throws an error 
![image](https://github.com/lightdash/lightdash/assets/1983672/97df9fde-4de3-4609-9164-7a8c5b295370)

Shifan-Gu commented:
Hi @rephus, 

thanks for the feedback! i will work on the comments.

rephus commented:
Also, some minor things (these can be addressed in separate issues) 

- We're missing the `filter by` options we have in results table 
![Screenshot from 2023-11-20 10-37-08](https://github.com/lightdash/lightdash/assets/1983672/6163ea65-c763-4ca1-95c5-046c6f4a9417)
![Screenshot from 2023-11-20 10-37-12](https://github.com/lightdash/lightdash/assets/1983672/ce212979-07e1-47c7-a565-94a097a526dc)

- these filters don't work on dashboards 

![image](https://github.com/lightdash/lightdash/assets/1983672/47b451a3-0f06-44a9-8964-71d3071abfd8)

rephus commented:
> Using a more complex TC expression (just adding brackets) does not work . It is silently not filtering anything.

Could we throw an exception (eg: not supported) every time a Table calculation filter is not supported ? rather than silently failing with a `1==1` condition .

Shifan-Gu commented:
Hi @rephus 

I was changing DEFAULT type table calculation filter to be used as string filter, but i noticed that for dimension and metrics they use <FilterStringAutoComplete/> which fetches the data from backend to pre-populate the select dropdown list.

Table calculation column is derived column so it can't have the same behaviour. i was thinking using the TagsInput instead of MultiSelect input but it requires the Maintine version 7.2.2.

Do you guys have any plan to upgrade Maintine to the latest version or do you have any suggestion on what component i can use for this filter?

Shifan-Gu commented:
> Returning a string (concat) throws an error ![image](https://user-images.githubusercontent.com/1983672/284223478-97df9fde-4de3-4609-9164-7a8c5b295370.png)

Checked this issue and i think its an valid error, because the table calculation result is cast to string but then its trying to filter as numeric value. 

You will see the same type of error when you creating the table calculation with string return type but set the format to number

Shifan-Gu commented:
> Using a more complex TC expression (just adding brackets) does not work . It is silently not filtering anything.
> 
> ![Screenshot from 2023-11-20 10-39-15](https://user-images.githubusercontent.com/1983672/284222887-a9e0334a-aa0c-4a88-a831-468184c6bfbf.png)
> 
> ![Screenshot from 2023-11-20 10-39-08](https://user-images.githubusercontent.com/1983672/284222898-5dab4e57-8077-46c6-b6e1-1438ec5d9822.png)
> 
> ![Screenshot from 2023-11-20 10-38-59](https://user-images.githubusercontent.com/1983672/284222907-38823079-3b0c-4e9e-b697-c7bfe7c0269e.png)

I checked this issue as well, its because when you edit the table calculation in the modal if you don't give it a new name, it will call a method to make it unique, so it's an expected behaviour.

So if you delete that filter and create it again it should work

Shifan-Gu commented:
Hi team, new code patch is ready for review

updates: 
1. resolved conflicts and fixed build issues
2. fixed broken e2e api tests
3. changed table calculation with Default format to use string filter component
4. throws the exception when a table calculation format type is not supported to render a sql

rephus commented:
Something weird is happening on the filter, the "field" thinks my table calculation is a number (see icon) but the filters I get are for `strings` (I should get `greater than`, etc) 

![image](https://github.com/lightdash/lightdash/assets/1983672/a8b4445e-e77b-4921-9fa2-b8cf9146fbef)

rephus commented:
Every time I update the table calculation, I loose the filter, it seems the table calculation id is changing. This can be fixed in a separate PR , https://github.com/lightdash/lightdash/issues/7996



[Screencast from 22-11-23 09:17:55.webm](https://github.com/lightdash/lightdash/assets/1983672/c2e4799c-33c1-4f4d-921d-47691663e246)

rephus commented:
Edge case, if I try to filter on a `date` field, I get this error, this can be fixed in a separate PR 

![image](https://github.com/lightdash/lightdash/assets/1983672/a369ad94-3ab8-46e9-b1b3-a3513bf4a09c)

rephus reviewed (approved):
The filter functionality works as expected, great job 

There are a few minor issues with types on table calculations but they can be fixed separately.
